### ISSUE TITLE
com.github.fommil.netlib:all:jar:1.1.2

### ISSUE THREAD

**Issue by Vzzarr** [broadinstitute/gatk#3724]
I am starting a Maven Project in which I would like to import your library; so I added this [dependency](http://search.maven.org/#artifactdetails%7Corg.broadinstitute%7Cgatk%7C4.beta.2%7C) to my pom.xml
```
<dependency>
    <groupId>org.broadinstitute</groupId>
    <artifactId>gatk</artifactId>
    <version>4.beta.2</version>
</dependency>
```
When I execute `mvn clear install` in my folder project, I receive this error: 
```
[ERROR] Failed to execute goal on project GATKpipe: 
Could not resolve dependencies for project uk.ac.ncl:GATKpipe:jar:0.0.1-SNAPSHOT: 
Could not find artifact com.github.fommil.netlib:all:jar:1.1.2 in 
all (https://mvnrepository.com/artifact/com.github.fommil.netlib/all) -> [Help 1]
```
and it seems that the problem is the dependency by com.github.fommil.netlib/all, indeed according to the output of `mvn clear install`, it attempt to download all-1.1.2.jar:
`Downloading: https://repo.maven.apache.org/maven2/com/github/fommil/netlib/all/1.1.2/all-1.1.2.jar`
but this jar is not available in the repository.

I noticed that even in other [projects](https://github.com/amplab/ml-matrix/issues/11) have the same issue.

How is possible to resolve this issue? 

Thanks for your time,
Nicholas

**Comment by lbergelson** [comment]
Hi @Vzzarr ,

Thank you for reporting this.  I've never heard of this happening before, but I suspect the difference is that you are using maven and we've only ever used gradle.  @magicDGS @LeeTL1220 You both build downstream projects, have you ever seen this?

From what I can tell `com.github.fommil.netlib.all` is a POM only dependency, but it's not annotated as such in our generated POM.  Gradle resolves it correctly, and then resolves it's transitive dependencies correctly, but maven seems to resolve it differently.  

I've reproduced the issue locally, and it seems like it can be fixed by changing this line in our build file:

https://github.com/broadinstitute/gatk/blob/62e339fb9150db19f80d4b48aed8d47608461690/build.gradle#L220

to specify the exact dependencies we want instead of using the all artifact.

I'll open a pull request with that.  There's an additional catch though, you're using the latest release to maven central, which is actually a very old version.  We haven't been able to release newer versions to central because they depend on a snapshot of a fork of a google library that isn't released to central.  We publish these to our artifactory server which can be resolved like any other maven repository.   Would you be able to work with an artifactory dependency?  If you can you can A) use the newer and better gatk, B) I won't have to backport this change on top of beta.2 

We're going to fix the problem of not being in central eventually ( certainly before 4.0 release... ) but until then resolving from our artifactory is probably the best bet.  If you can't / don't want to use the artifactory releases, let me know and I can release a patch version to maven that fixes this problem.

**Comment by lbergelson** [comment]
@Vzzarr I believe this is fixed in the most recent master now.  Could you try building with the version published to our [artifactory](
https://broadinstitute.jfrog.io/broadinstitute/webapp/#/artifacts/browse/tree/General/libs-snapshot-local/org/broadinstitute/gatk/4.beta.6-18-g2ee7724-SNAPSHOT/gatk-4.beta.6-18-g2ee7724-20171025.162137-1.jar) to confirm that the fix works?

**Comment by Vzzarr** [comment]
Hi @lbergelson and thanks for considering my issue,

I'm sorry but I'm not familiar to artifactory dependency, if necessary I'll deepen about it; so I just inserted this dependency in the project's pom.xml
```
<dependency>
    <groupId>org.broadinstitute</groupId>
    <artifactId>gatk</artifactId>
    <version>4.beta.6-18-g2ee7724-20171025.162137-1</version>
</dependency>
```
as reported in the [artifact repository](https://broadinstitute.jfrog.io/broadinstitute/webapp/#/artifacts/browse/tree/General/libs-snapshot-local/org/broadinstitute/gatk/4.beta.6-18-g2ee7724-SNAPSHOT/gatk-4.beta.6-18-g2ee7724-20171025.162137-1.jar), but when I execute `mvn clear install` in my folder project, I receive this error: 
```
[ERROR] Failed to execute goal on project GATKpipe: 
Could not resolve dependencies for project uk.ac.ncl:GATKpipe:jar:0.0.1-SNAPSHOT: 
Could not find artifact org.broadinstitute:gatk:jar:4.beta.6-18-g2ee7724-20171025.162137-1 -> [Help 1]
```

Am I making any mistake?

**Comment by lbergelson** [comment]
@Vzzarr Ah, sorry, I should have explained.  It's a different repository from maven central, so you'll have to add an extra repository resolution block to your pom.

 ```
    <repositories>
       <repository>
          <snapshots />
          <id>snapshots</id>
          <name>libs-snapshot</name>
          <url>https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot</url>
        </repository>
    </repositories>
```

I think this should work, but I'm very much a maven novice, so you may have to tweak something. This will ask it to resolve artifacts from the Broad Institute's artifactory which includes our gatk snapshots.  

If you're not attached to maven as your build system, we definitely recommend trying out gradle. I like it a lot more than maven, although it definitely has it's own complications.

**Comment by Vzzarr** [comment]
@lbergelson I used Maven because it's what they taught me at the university to download Spark dependencies, so I'm a novice too; I'm sorry to bothering you with new errors (I promise that tomorrow I see how to do it with Gradle and stop bothering you), but adding even the repository to the pom.xml generated this error: 
```
[ERROR] Failed to execute goal on project GATKpipe: 
Could not resolve dependencies for project uk.ac.ncl:GATKpipe:jar:0.0.1-SNAPSHOT: 
Failed to collect dependencies at org.broadinstitute:gatk:jar:4.beta.6-18-g2ee7724-20171025.162137-1 
-> com.google.cloud.genomics:google-genomics-dataflow:jar:v1beta2-0.15 
-> com.google.cloud.genomics:google-genomics-utils:jar:v1beta2-0.30 
-> io.grpc:grpc-all:jar:0.7.1 -> io.grpc:grpc-auth:jar:0.7.1 
-> com.google.auth:google-auth-library-oauth2-http:jar:0.1.0: 
Failed to read artifact descriptor for com.google.auth:google-auth-library-oauth2-http:jar:0.1.0: 
Could not find artifact com.google.auth:google-auth-library:pom:0.1.0 in snapshots (https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot) -> [Help 1]
```

**Comment by lbergelson** [comment]
@Vzzarr Don't worry, these things are complicated and take time to learn!  You can use whatever you want, I just know more gradle than maven so it's easier for me to help with that.  

I'm having trouble reproducing the error you're having.  Is your project on github? Or at a minimum could you paste your pom file here?  

In the meantime, could you try cloning https://github.com/lbergelson/gatk-downstream-test and seeing if `mvn compile` completes successfully?

**Comment by Vzzarr** [comment]
I just created a Maven Project with Eclipse and modified the pom.xml as reported [here](https://pastebin.com/kwi7gSRk)
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>uk.ac.ncl</groupId>
    <artifactId>GATKpipe</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <repositories>
        <repository>
            <snapshots />
            <id>snapshots</id>
            <name>libs-snapshot</name>
            <url>https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot</url>
        </repository>
    </repositories>
    <dependencies>
        <dependency>
            <groupId>org.broadinstitute</groupId>
            <artifactId>gatk</artifactId>
            <version>4.beta.6-18-g2ee7724-20171025.162137-1</version>
        </dependency>
    </dependencies>
</project>
```

**Comment by lbergelson** [comment]
I think there's some sort of permissions issue here.  I tried building while not connected to our internal network and ran into a similar issue.  I don't exactly understand how our artifactory serves as a mirror of central, but it seems like it may not be mirroring correctly to people outside of our network. 

@davidbernick it seems like external people can resolve the artifacts we host in libs-snapshot-local, but possibly not things in libs-snapshot and libs-release.  Does that make any sense to you?

**Comment by magicDGS** [comment]
My software is correctly pulling correctly the artifacts in https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot/, and I am not at Broad...

**Comment by lbergelson** [comment]
@magicdgs Do you get all your dependencies from there, or only gatk?  I assumed you had multiple repos configured so that it first resolves from central and then from artifactory.  I'm not sure how to configure that sanely in maven, although I'm sure it's doable to someone who is less ignorant of maven then me.

**Comment by magicDGS** [comment]
@lbergelson  - In gradle, I first resolve with maven central and then with your artifactory:

```gradle
repositories {
    mavenCentral()
    maven {
        url "https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot/"
    }
}
```

In the case of maven, for several repositories this should be done following [this](https://maven.apache.org/guides/mini/guide-multiple-repositories.html). I think that the configuration for the repositories should look like this (if I remember correctly):

```xml
<repositories>
    <repository>
        <id>central</id>
        <name>Maven Repository Switchboard</name>
        <layout>default</layout>
        <url>http://repo1.maven.org/maven2</url>
        <snapshots>
            <enabled>false</enabled>
        </snapshots>
    </repository>
    <repository>
        <id>snapshots</id>
        <snapshots>
            <enabled>true</enabled>
        </snapshots>
        <name>libs-snapshot</name>
        <url>https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot</url>
    </repository>
</repositories>
```

**Comment by lbergelson** [comment]
Oh, I didn't realize you could just specify multiple different repos in maven like that as well.  That makes total sense.  @Vzzarr I would try that.

**Comment by Vzzarr** [comment]
I deleted ~/.m2/repository/ folder and his content, used this [pom.xml](https://pastebin.com/mV1qGTzv)
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>uk.ac.ncl</groupId>
    <artifactId>GATKpipe</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <repositories>
        <repository>
            <id>central</id>
            <name>Maven Repository Switchboard</name>
            <layout>default</layout>
            <url>http://repo1.maven.org/maven2</url>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
        </repository>
        <repository>
            <id>snapshots</id>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
            <name>libs-snapshot</name>
            <url>https://broadinstitute.jfrog.io/broadinstitute/libs-snapshot</url>
        </repository>
    </repositories>
    <dependencies>
        <dependency>
            <groupId>org.broadinstitute</groupId>
            <artifactId>gatk</artifactId>
            <version>4.beta.6-18-g2ee7724-20171025.162137-1</version>
        </dependency>
    </dependencies>
</project>
```
executed `mvn clean install -U` and finally obtained the `BUILD SUCCESS`
Thanks everybody for your support

**Comment by lbergelson** [comment]
Hooray!  Sorry this was so painful.  Closing this ticket, feel free to open a new one if you run into other issues.

**Comment by lbergelson** [comment]
Fixed by #3742
